File: tests/modules/good/c/text.c
/* This is a template file following the expected interface and declarations to
 * implement the corresponding Catala module.
 *
 * You should replace all `catala_error(catala_impossible)` place-holders with
 * your implementation and rename it to remove the ".template" suffix. */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <catala_runtime.h>

#include <Stdlib_en.h>
#include <Date_en.h>
#include <List_en.h>
#include <Duration_en.h>
#include <MonthYear_en.h>
#include <Period_en.h>
#include <Money_en.h>
#include <Integer_en.h>
#include <Decimal_en.h>

typedef const char* Text_Text;

int catala_type__Text_Text_equal(const catala_code_position* pos,
                                     const Text_Text t1,
                                     const Text_Text t2) {
  if (strcmp(t1, t2) == 0) return 1;
  else return 0;
}
int catala_type__Text_Text_compare(const catala_code_position* pos,
                                       const Text_Text t1,
                                       const Text_Text t2) {
  int c = strcmp(t1, t2);
  return c > 0 ? 1 : c < 0 ? -1 : 0;
}
void catala_type__Text_Text_print(struct catala_buf buf, const Text_Text t) {
  buf.printf("\"%s\"", t);
}
void catala_type__Text_Text_to_json(struct catala_buf buf, const Text_Text t) {
  int i = 0;
  buf.printf("\"");
  while (1) {
    char c = t[i++];
    switch (c) {
    case '\0': buf.printf("\""); return;
    case '\"': case '\\': buf.printf("\\%c", c); break;
    case '\b': buf.printf("\\b"); break;
    case '\f': buf.printf("\\f"); break;
    case '\n': buf.printf("\\n"); break;
    case '\r': buf.printf("\\r"); break;
    case '\t': buf.printf("\\t"); break;
    default: buf.printf("%c", c);
    }
  }
}

char read_hex_char(const catala_code_position* pos, char c) {
  if ('0' <= c && c <= '9') return c - '0';
  if ('a' <= c && c <= 'f') return 10 + c - 'a';
  if ('A' <= c && c <= 'F') return 10 + c - 'A';
  catala_error(catala_assertion_failed, pos, 1, "Invalid JSON input for type text");
  abort();
}

const Text_Text catala_type__Text_Text_from_json(const catala_code_position* pos, const char * t) {
  int len = strlen(t);
  char* buf;
  int i, ofs;
  if (len < 2 || t[0] != '"' || t[len - 1] != '"')
    catala_error(catala_assertion_failed, pos, 1, "Invalid JSON input for type text");
  buf = catala_malloc (strlen(t) - 2 + 1);
  for (i = 0, ofs = 1; ofs < len - 1; i++, ofs++) {
    switch (t[ofs]) {
    case '"':
      catala_error(catala_assertion_failed, pos, 1, "Invalid JSON input for type text");
    case '\\':
      ofs++;
      if (ofs >= len - 1)
        catala_error(catala_assertion_failed, pos, 1, "Invalid JSON input for type text");
      switch (t[ofs]) {
      case 'b': buf[i] = '\b'; break;
      case 'f': buf[i] = '\f'; break;
      case 'n': buf[i] = '\n'; break;
      case 'r': buf[i] = '\r'; break;
      case 't': buf[i] = '\t'; break;
      case '"': case '/': case '\\':
        buf[i] = t[ofs]; break;
      case 'u':
        if (ofs >= len - 5)
          catala_error(catala_assertion_failed, pos, 1, "Invalid JSON input for type text");
        else {
          char w = read_hex_char(pos, t[++ofs]);
          char x = read_hex_char(pos, t[++ofs]);
          char y = read_hex_char(pos, t[++ofs]);
          char z = read_hex_char(pos, t[++ofs]);
          if (w == 0 && x == 0 && y < 8) { /* 1-byte encoding (ASCII) */
            buf[i] = y << 4 | z;
          }
          else if (x < 8) { /* 2-byte encoding */
            buf[i] = 0xc0 | x << 2 | y >> 2;
            buf[++i] = 0x80 | (y << 6 >> 2) | z;
          }
          else  { /* 3-byte encoding */
            buf[i] = 0xe0 | w;
            buf[++i] = 0x80 | (x << 2) | (y >> 2);
            buf[++i] = 0x80 | (y << 6 >> 2) | z;
          }
        }
        break;
      default:
        catala_error(catala_assertion_failed, pos, 1, "Invalid JSON input for type text");
      }
      break;
    default:
      buf[i] = t[ofs];
    }
  }
  buf[i] = '\0';
  return buf;
}
const catala_type catala_type__Text_Text()
{
  static catala_type ty = {UNINITIALIZED};
  if (ty.kind != UNINITIALIZED) return ty;
  ty.contents.texternal.name = "Text_Text";
  ty.contents.texternal.equal =
    (int (*)(const catala_code_position *, const void *, const void *))
    &catala_type__Text_Text_equal;
  ty.contents.texternal.compare =
    (int (*)(const catala_code_position *, const void *, const void *))
    &catala_type__Text_Text_compare;
  ty.contents.texternal.print =
    (void (*)(struct catala_buf, const void *))
    &catala_type__Text_Text_print;
  ty.contents.texternal.to_json =
    (void (*)(struct catala_buf, const void *))
    &catala_type__Text_Text_to_json;
  ty.contents.texternal.from_json =
    (void * (*)(const catala_code_position *, const char *))
    &catala_type__Text_Text_from_json;
  ty.kind = EXTERNAL;
  return ty;
}

const catala_code_position Text__loc[][1] = {
  {{"tests/modules/good/text.catala_en", 8, 13, 8, 16}},
  {{"tests/modules/good/text.catala_en", 9, 13, 9, 16}},
  {{"tests/modules/good/text.catala_en", 10, 13, 10, 21}},
  {{"tests/modules/good/text.catala_en", 11, 13, 11, 19}}
};

const Text_Text Text__foo () {
  return "foo\\";
}

const Text_Text Text__bar () {
  return "bąr";
}

const Text_Text Text__fortytwo () {
  return "42";
}

const Text_Text Text__of_int (CATALA_INT x)
{
  catala_print(catala_strbuf, embed(catala_type_integer, x));
  return catala_strbuf.flush();
}

File: tests/modules/good/ocaml/text.ml
(* This is a template file following the expected interface and declarations to
 * implement the corresponding Catala module.
 *
 * You should replace all `raise (Error (Impossible))` place-holders with your
 * implementation and rename it to remove the ".template" suffix. *)

[@@@ocaml.warning "-4-26-27-32-33-34-37-41-42-69"]

open Catala_runtime

module Text = ExternalType(struct
    module String = Stdlib.String
    type t = string
    let name = "Text"
    let equal _pos t1 t2 = String.equal t1 t2
    let compare _pos t1 t2 = String.compare t1 t2
    let print t = Printf.sprintf "%S" t
    let to_json t =
      let buf = Buffer.create ((2 * String.length t) + 2) in
      Buffer.add_char buf '"';
      String.iter
        (function
          | ('"' | '\\') as c ->
            Buffer.add_char buf '\\';
            Buffer.add_char buf c
          | '\b' -> Buffer.add_string buf "\\b"
          | '\x0c' -> Buffer.add_string buf "\\f"
          | '\n' -> Buffer.add_string buf "\\n"
          | '\r' -> Buffer.add_string buf "\\r"
          | '\t' -> Buffer.add_string buf "\\t"
          | c -> Buffer.add_char buf c)
        t;
      Buffer.add_char buf '"';
      Buffer.contents buf
    let from_json pos s =
      let err() =
        raise (Error (AssertionFailed, [pos], Some "Invalid JSON input for type string"))
      in
      let len = String.length s in
      if len < 2 || s.[0] <> '"' || s.[len - 1] <> '"' then err();
      let buf = Buffer.create len in
      let rec iter i =
        if i < len - 1 then
          match s.[i] with
          | '\\' ->
            let i = i + 1 in
            if i >= len - 1 then err();
            (match s.[i] with
             | 'b' -> Buffer.add_char buf '\b'; iter (i+1)
             | 'f' -> Buffer.add_char buf '\x0c'; iter (i+1)
             | 'n' -> Buffer.add_char buf '\n'; iter (i+1)
             | 'r' -> Buffer.add_char buf '\r'; iter (i+1)
             | 't' -> Buffer.add_char buf '\t'; iter (i+1)
             | '"' | '/' | '\\' as c -> Buffer.add_char buf c; iter (i+1)
             | 'u' ->
               if i >= len - 5 then err ();
               let () =
                 match int_of_string_opt ("0x" ^ String.sub s (i+1) 4) with
                 | None -> err()
                 | Some n ->
                   try Buffer.add_utf_8_uchar buf (Uchar.of_int n)
                   with Invalid_argument _ -> err()
               in
               iter (i + 4)
             | _ -> err())
          | c -> Buffer.add_char buf c; iter (i+1)
      in
      iter 1;
      Buffer.contents buf
end)

(* Toplevel def loc *)
let loc : code_location array =
  [|{filename="tests/modules/good/text.catala_en";
     start_line=8; start_column=13; end_line=8; end_column=16;
     law_headings=["Text types in Catala"]};
    {filename="tests/modules/good/text.catala_en";
     start_line=9; start_column=13; end_line=9; end_column=16;
     law_headings=["Text types in Catala"]};
    {filename="tests/modules/good/text.catala_en";
     start_line=10; start_column=13; end_line=10; end_column=19;
     law_headings=["Text types in Catala"]}|]

(* Toplevel def foo *)
let foo : Text.t = "foo\\"

(* Toplevel def bar *)
let bar : Text.t = "bąr"

let fortytwo = "42"

(* Toplevel def of_int *)
let of_int : integer -> Text.t = integer_to_string

let () =
  Catala_runtime.register_module "Text"
    [ "foo", Stdlib.Obj.repr (foo);
      "bar", Stdlib.Obj.repr (bar);
      "fortytwo", Stdlib.Obj.repr (fortytwo);
      "of_int", Stdlib.Obj.repr (of_int) ]
    ~types:["Text", (module Text)]
    "*external*"
